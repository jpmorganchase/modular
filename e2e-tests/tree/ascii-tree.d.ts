declare module 'ascii-tree' {
  export function generate(input: string): string;
}
