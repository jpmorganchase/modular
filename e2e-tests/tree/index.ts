import fs from 'fs';
import path from 'path';
import asciiTree from 'ascii-tree';
import hash from '@emotion/hash';

function times(str: string, length: number) {
  return Array.from({ length }, () => str).join('');
}

const defaultIgnores = ['node_modules', '.git', '.DS_Store'];
const defaultHashIgnores = [
  // adding lockfiles here because it can be different
  // on different runs; since we install the latest versions
  // of some packages when making a repository
  'yarn.lock',
  'package-lock.json',
  // adding package.json files since versions within these get bumped
  'package.json',
];

function tree(
  _path: string,
  level = 1,
  options = {
    ignores: defaultIgnores,
    hashIgnores: defaultHashIgnores,
  },
): string {
  const stat = fs.statSync(_path);
  if (stat.isDirectory()) {
    const children = fs.readdirSync(_path);
    const dirArr = _path.split('/');
    const dir = dirArr[dirArr.length - 1];
    // todo - should these be sorted?
    // todo - handle symlinks, etc
    return `${times('#', level)}${dir}\n${children
      .filter((child: string) => !options.ignores.includes(child))
      .map((child: string) => tree(path.join(_path, child), level + 1, options))
      .join('\n')}`;
  } else {
    return [
      `${times('#', level)}${path.basename(_path)}`,
      options.hashIgnores.includes(path.basename(_path))
        ? undefined
        : `#${hash(fs.readFileSync(_path, 'utf8'))}`,
    ]
      .filter(Boolean)
      .join(' ');
  }
}

export default function generate(_path: string): string {
  return asciiTree.generate(tree(_path));
}
