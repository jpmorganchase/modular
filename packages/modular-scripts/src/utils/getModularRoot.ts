import * as path from 'path';
import * as fs from 'fs-extra';
import findUp from 'find-up';

function isModularRoot(packageJson: { modular?: Record<string, unknown> }) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory: string) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

export default function getModularRoot(): string {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
}
