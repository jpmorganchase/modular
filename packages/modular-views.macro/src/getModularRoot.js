// keep this in sync with modular-scripts/src/getModularRoot
const path = require('path');
const fs = require('fs-extra');
const findUp = require('find-up');

function isModularRoot(packageJson) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

module.exports = function getModularRoot() {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
};
