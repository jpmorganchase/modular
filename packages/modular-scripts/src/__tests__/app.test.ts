import execa from 'execa';
import { exec } from 'child_process';
import rimraf from 'rimraf';
import tree from 'tree-view-for-tests';
import path from 'path';
import fs from 'fs-extra';
import prettier from 'prettier';

import {
  getDocument,
  getQueriesForElement,
  queries,
} from 'pptr-testing-library';
import getModularRoot from '../utils/getModularRoot';
import puppeteer from 'puppeteer';

import { startApp, DevServer } from './start-app';
import type { CoreProperties } from '@schemastore/package';

// eslint-disable-next-line @typescript-eslint/unbound-method
const { getNodeText } = queries;

const modularRoot = getModularRoot();

// These tests must be executed sequentially with `--runInBand`.

const packagesPath = path.join(getModularRoot(), 'packages');

function modular(str: string, opts: Record<string, unknown> = {}) {
  return execa('yarnpkg', ['modular', ...str.split(' ')], {
    cwd: modularRoot,
    cleanup: true,
    ...opts,
  });
}

function cleanup() {
  rimraf.sync(path.join(packagesPath, 'sample-app'));
  rimraf.sync(path.join(modularRoot, 'dist/sample-app'));

  rimraf.sync(path.join(packagesPath, 'scoped/sample-app'));
  rimraf.sync(path.join(modularRoot, 'dist/scoped-sample-app'));

  rimraf.sync(path.join(packagesPath, 'node-env-app'));
  rimraf.sync(path.join(modularRoot, 'dist/node-env-app'));

  // run yarn so yarn.lock gets reset
  return execa.sync('yarnpkg', ['--silent'], {
    cwd: modularRoot,
  });
}

beforeAll(cleanup);
afterAll(cleanup);

describe('when working with a NODE_ENV app', () => {
  beforeAll(async () => {
    await modular(
      'add node-env-app --unstable-type app --unstable-name node-env-app',
      { stdio: 'inherit' },
    );

    await fs.writeFile(
      path.join(modularRoot, 'packages', 'node-env-app', 'src', 'index.ts'),
      `
      console.log(process.env.NODE_ENV);

      export {};
    `,
    );

    await modular('build node-env-app', {
      stdio: 'inherit',
    });
  });

  it('can build a app', () => {
    expect(tree(path.join(modularRoot, 'dist', 'node-env-app')))
      .toMatchInlineSnapshot(`
      "node-env-app
      ├─ asset-manifest.json #u3l86j
      ├─ favicon.ico #6pu3rg
      ├─ index.html #18m4pzt
      ├─ logo192.png #1nez7vk
      ├─ logo512.png #1hwqvcc
      ├─ manifest.json #19gah8o
      ├─ package.json
      ├─ robots.txt #1sjb8b3
      └─ static
         └─ js
            ├─ main.6b9113da.js #1jh9hcj
            ├─ main.6b9113da.js.map #2l1ifi
            ├─ runtime-main.182069c4.js #r0hm4v
            └─ runtime-main.182069c4.js.map #ul1xez"
    `);
  });

  it('can generate a js/main.6b9113da.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'node-env-app',
              'static',
              'js',
              'main.6b9113da.js',
            ),
          ),
        ),
        {
          filepath: 'main.6b9113da.js',
        },
      ),
    ).toMatchSnapshot();
  });
});

describe('When working with a nested app', () => {
  beforeAll(async () => {
    await modular(
      'add scoped/sample-app --unstable-type app --unstable-name @scoped/sample-app',
      { stdio: 'inherit' },
    );

    await modular('build @scoped/sample-app', {
      stdio: 'inherit',
    });
  });

  it('can build a nested app', () => {
    expect(tree(path.join(modularRoot, 'dist', 'scoped-sample-app')))
      .toMatchInlineSnapshot(`
      "scoped-sample-app
      ├─ asset-manifest.json #mfwa5x
      ├─ favicon.ico #6pu3rg
      ├─ index.html #m0ahw0
      ├─ logo192.png #1nez7vk
      ├─ logo512.png #1hwqvcc
      ├─ manifest.json #19gah8o
      ├─ package.json
      ├─ robots.txt #1sjb8b3
      └─ static
         ├─ css
         │  ├─ main.1a7488ce.css #x701i6
         │  └─ main.1a7488ce.css.map #z36y5v
         └─ js
            ├─ 788.78cfb599.js #1bgvsgc
            ├─ 788.78cfb599.js.LICENSE.txt #eplx8h
            ├─ 788.78cfb599.js.map #jf9876
            ├─ main.8c5bc4d2.js #xquhbv
            ├─ main.8c5bc4d2.js.map #t8irmb
            ├─ runtime-main.83e93d07.js #1mo3tl5
            └─ runtime-main.83e93d07.js.map #3ycdm3"
    `);
  });

  it('can generate a asset-manifest', async () => {
    expect(
      String(
        await fs.readFile(
          path.join(
            modularRoot,
            'dist',
            'scoped-sample-app',
            'asset-manifest.json',
          ),
        ),
      ),
    ).toMatchSnapshot();
  });

  it('can generate a manifest', async () => {
    expect(
      String(
        await fs.readFile(
          path.join(modularRoot, 'dist', 'scoped-sample-app', 'manifest.json'),
        ),
      ),
    ).toMatchSnapshot();
  });

  it('can generate a index.html', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(modularRoot, 'dist', 'scoped-sample-app', 'index.html'),
          ),
        ),
        {
          filepath: 'index.html',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a css/main.1a7488ce.css', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'scoped-sample-app',
              'static',
              'css',
              'main.1a7488ce.css',
            ),
          ),
        ),
        {
          filepath: 'main.1a7488ce.css',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a css/main.1a7488ce.css.map', async () => {
    expect(
      JSON.parse(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'scoped-sample-app',
              'static',
              'css',
              'main.1a7488ce.css.map',
            ),
          ),
        ),
      ),
    ).toMatchSnapshot();
  });

  it('can generate a js/main.8c5bc4d2.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'scoped-sample-app',
              'static',
              'js',
              'main.8c5bc4d2.js',
            ),
          ),
        ),
        {
          filepath: 'main.8c5bc4d2.js',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a js/runtime-main.83e93d07.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'scoped-sample-app',
              'static',
              'js',
              'runtime-main.83e93d07.js',
            ),
          ),
        ),
        {
          filepath: 'runtime-main.83e93d07.js',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a js/788.78cfb599.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'scoped-sample-app',
              'static',
              'js',
              '788.78cfb599.js',
            ),
          ),
        ),
        {
          filepath: '788.78cfb599.js',
        },
      ),
    ).toMatchSnapshot();
  });
});

describe('when working with an app', () => {
  beforeAll(async () => {
    await modular(
      'add sample-app --unstable-type app --unstable-name sample-app',
      { stdio: 'inherit' },
    );

    // Let's replace the App module with something of our own
    // with a test specific element we can introspect
    await fs.copyFile(
      path.join(__dirname, 'TestApp.test-tsx'),
      path.join(packagesPath, 'sample-app', 'src', 'App.tsx'),
    );

    await modular('build sample-app', {
      stdio: 'inherit',
    });
  });

  it('can add an app', () => {
    expect(tree(path.join(packagesPath, 'sample-app'))).toMatchInlineSnapshot(`
      "sample-app
      ├─ package.json
      ├─ public
      │  ├─ favicon.ico #6pu3rg
      │  ├─ index.html #1m6toxd
      │  ├─ logo192.png #1nez7vk
      │  ├─ logo512.png #1hwqvcc
      │  ├─ manifest.json #19gah8o
      │  └─ robots.txt #1sjb8b3
      ├─ src
      │  ├─ App.css #1o0zosm
      │  ├─ App.tsx #igvgtx
      │  ├─ __tests__
      │  │  └─ App.test.tsx #16urcos
      │  ├─ index.css #o7sk21
      │  ├─ index.tsx #zdn6mw
      │  ├─ logo.svg #1okqmlj
      │  └─ react-app-env.d.ts #t4ygcy
      └─ tsconfig.json #6rw46b"
    `);
  });

  it('can build an app', () => {
    expect(tree(path.join(modularRoot, 'dist', 'sample-app')))
      .toMatchInlineSnapshot(`
      "sample-app
      ├─ asset-manifest.json #nzl5za
      ├─ favicon.ico #6pu3rg
      ├─ index.html #11ep3dl
      ├─ logo192.png #1nez7vk
      ├─ logo512.png #1hwqvcc
      ├─ manifest.json #19gah8o
      ├─ package.json
      ├─ robots.txt #1sjb8b3
      └─ static
         ├─ css
         │  ├─ main.1a7488ce.css #x701i6
         │  └─ main.1a7488ce.css.map #z36y5v
         └─ js
            ├─ 316.7a4d5eb7.js #1yh1m0p
            ├─ 316.7a4d5eb7.js.LICENSE.txt #eplx8h
            ├─ 316.7a4d5eb7.js.map #1ib48x2
            ├─ main.14cec3f9.js #12mskyj
            ├─ main.14cec3f9.js.map #1x1swby
            ├─ runtime-main.de9fd74d.js #1slkp3g
            └─ runtime-main.de9fd74d.js.map #d7r4im"
    `);
  });

  it('can generate a asset-manifest', async () => {
    expect(
      String(
        await fs.readFile(
          path.join(modularRoot, 'dist', 'sample-app', 'asset-manifest.json'),
        ),
      ),
    ).toMatchSnapshot();
  });

  it('can generate a manifest', async () => {
    expect(
      String(
        await fs.readFile(
          path.join(modularRoot, 'dist', 'sample-app', 'manifest.json'),
        ),
      ),
    ).toMatchSnapshot();
  });

  it('can generate a package.json', async () => {
    const packageJson = JSON.parse(
      String(
        await fs.readFile(
          path.join(modularRoot, 'dist', 'sample-app', 'package.json'),
        ),
      ),
    ) as CoreProperties;

    expect(packageJson.name).toBe('sample-app');
    expect(packageJson.version).toBe('0.1.0');
    expect(packageJson.modular).toStrictEqual({ type: 'app' });
    expect(packageJson.dependencies?.react).toBeTruthy();
    expect(packageJson.dependencies?.['react-dom']).toBeTruthy();
  });

  it('can generate a index.html', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(modularRoot, 'dist', 'sample-app', 'index.html'),
          ),
        ),
        {
          filepath: 'index.html',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a js/main.14cec3f9.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'sample-app',
              'static',
              'js',
              'main.14cec3f9.js',
            ),
          ),
        ),
        {
          filepath: 'main.14cec3f9.js',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a js/runtime-main.de9fd74d.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'sample-app',
              'static',
              'js',
              'runtime-main.de9fd74d.js',
            ),
          ),
        ),
        {
          filepath: 'runtime-main.de9fd74d.js',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can generate a js/316.7a4d5eb7.js', async () => {
    expect(
      prettier.format(
        String(
          await fs.readFile(
            path.join(
              modularRoot,
              'dist',
              'sample-app',
              'static',
              'js',
              '316.7a4d5eb7.js',
            ),
          ),
        ),
        {
          filepath: '316.7a4d5eb7.js',
        },
      ),
    ).toMatchSnapshot();
  });

  it('can execute tests', async () => {
    const output = await modular('test sample-app --watchAll false', {
      all: true,
      reject: false,
      env: {
        CI: 'true',
      },
    });

    // TODO: Passing CI=true *should* remove all the coloring stuff,
    // it's weird that it doesn't. To workaround it, I've manually
    // removed those tokens from the string for the snapshot test.
    // Open to suggestions/fixes.

    // eslint-disable-next-line no-control-regex
    const cleanedOutput = output.all?.replace(/|\[\d+./gm, '');

    expect(cleanedOutput).toContain(
      'PASS test packages/sample-app/src/__tests__/App.test.tsx',
    );
  });

  it('can start an app', async () => {
    let browser: puppeteer.Browser | undefined;
    let devServer: DevServer | undefined;
    let port: string;
    try {
      await fs.copyFile(
        path.join(__dirname, 'TestApp.test-tsx'),
        path.join(packagesPath, 'sample-app', 'src', 'App.tsx'),
      );

      browser = await puppeteer.launch({
        // always run in headless - if you want to debug this locally use the env var to
        headless: !Boolean(process.env.NO_HEADLESS_TESTS),
        args: ['--no-sandbox', '--disable-setuid-sandbox'],
      });
      port = '3000';
      devServer = await startApp('sample-app', { env: { PORT: port } });

      const page = await browser.newPage();
      await page.goto(`http://localhost:${port}`, {});

      // eslint-disable-next-line @typescript-eslint/unbound-method
      const { getByTestId, findByTestId } = getQueriesForElement(
        await getDocument(page),
      );

      await findByTestId('test-this');

      const element = getByTestId('test-this');

      expect(await getNodeText(await element)).toBe('this is a modular app');
    } finally {
      if (browser) {
        await browser.close();
      }
      if (devServer) {
        // this is the problematic bit, it leaves hanging node processes
        // despite closing the parent process. Only happens in tests!
        devServer.kill();
      }
    }

    if (port) {
      // kill all processes listening to the dev server port
      exec(
        `lsof -n -i4TCP:${port} | grep LISTEN | awk '{ print $2 }' | xargs kill`,
        (err) => {
          if (err) {
            console.log('err: ', err);
          }
          console.log(`Cleaned up processes on port ${port}`);
        },
      );
    }
  });
});
