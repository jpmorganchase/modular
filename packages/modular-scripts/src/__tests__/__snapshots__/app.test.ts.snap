// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When working with a nested app can generate a asset-manifest 1`] = `
"{
  \\"files\\": {
    \\"main.css\\": \\"/static/css/main.f09a1402.chunk.css\\",
    \\"main.js\\": \\"/static/js/main.301a9b5d.chunk.js\\",
    \\"main.js.map\\": \\"/static/js/main.301a9b5d.chunk.js.map\\",
    \\"runtime-main.js\\": \\"/static/js/runtime-main.90523c3e.js\\",
    \\"runtime-main.js.map\\": \\"/static/js/runtime-main.90523c3e.js.map\\",
    \\"static/js/2.928f002b.chunk.js\\": \\"/static/js/2.928f002b.chunk.js\\",
    \\"static/js/2.928f002b.chunk.js.map\\": \\"/static/js/2.928f002b.chunk.js.map\\",
    \\"index.html\\": \\"/index.html\\",
    \\"static/css/main.f09a1402.chunk.css.map\\": \\"/static/css/main.f09a1402.chunk.css.map\\",
    \\"static/js/2.928f002b.chunk.js.LICENSE.txt\\": \\"/static/js/2.928f002b.chunk.js.LICENSE.txt\\"
  },
  \\"entrypoints\\": [
    \\"static/js/runtime-main.90523c3e.js\\",
    \\"static/js/2.928f002b.chunk.js\\",
    \\"static/css/main.f09a1402.chunk.css\\",
    \\"static/js/main.301a9b5d.chunk.js\\"
  ]
}"
`;

exports[`When working with a nested app can generate a index.html 1`] = `"<!doctype html><html lang=\\"en\\"><head><meta charset=\\"utf-8\\"/><link rel=\\"icon\\" href=\\"/favicon.ico\\"/><meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\"/><meta name=\\"theme-color\\" content=\\"#000000\\"/><meta name=\\"description\\" content=\\"Web site created using create-react-app\\"/><link rel=\\"apple-touch-icon\\" href=\\"/logo192.png\\"/><link rel=\\"manifest\\" href=\\"/manifest.json\\"/><title>React App</title><link href=\\"/static/css/main.f09a1402.chunk.css\\" rel=\\"stylesheet\\"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id=\\"root\\"></div><script>!function(e){function r(r){for(var n,p,l=r[0],a=r[1],f=r[2],c=0,s=[];c<l.length;c++)p=l[c],Object.prototype.hasOwnProperty.call(o,p)&&o[p]&&s.push(o[p][0]),o[p]=0;for(n in a)Object.prototype.hasOwnProperty.call(a,n)&&(e[n]=a[n]);for(i&&i(r);s.length;)s.shift()();return u.push.apply(u,f||[]),t()}function t(){for(var e,r=0;r<u.length;r++){for(var t=u[r],n=!0,l=1;l<t.length;l++){var a=t[l];0!==o[a]&&(n=!1)}n&&(u.splice(r--,1),e=p(p.s=t[0]))}return e}var n={},o={1:0},u=[];function p(r){if(n[r])return n[r].exports;var t=n[r]={i:r,l:!1,exports:{}};return e[r].call(t.exports,t,t.exports,p),t.l=!0,t.exports}p.m=e,p.c=n,p.d=function(e,r,t){p.o(e,r)||Object.defineProperty(e,r,{enumerable:!0,get:t})},p.r=function(e){\\"undefined\\"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:\\"Module\\"}),Object.defineProperty(e,\\"__esModule\\",{value:!0})},p.t=function(e,r){if(1&r&&(e=p(e)),8&r)return e;if(4&r&&\\"object\\"==typeof e&&e&&e.__esModule)return e;var t=Object.create(null);if(p.r(t),Object.defineProperty(t,\\"default\\",{enumerable:!0,value:e}),2&r&&\\"string\\"!=typeof e)for(var n in e)p.d(t,n,function(r){return e[r]}.bind(null,n));return t},p.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return p.d(r,\\"a\\",r),r},p.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},p.p=\\"/\\";var l=this[\\"webpackJsonp@scoped/sample-app\\"]=this[\\"webpackJsonp@scoped/sample-app\\"]||[],a=l.push.bind(l);l.push=r,l=l.slice();for(var f=0;f<l.length;f++)r(l[f]);var i=a;t()}([])</script><script src=\\"/static/js/2.928f002b.chunk.js\\"></script><script src=\\"/static/js/main.301a9b5d.chunk.js\\"></script></body></html>"`;

exports[`When working with a nested app can generate a manifest 1`] = `
"{
  \\"short_name\\": \\"React App\\",
  \\"name\\": \\"Create React App Sample\\",
  \\"icons\\": [
    {
      \\"src\\": \\"favicon.ico\\",
      \\"sizes\\": \\"64x64 32x32 24x24 16x16\\",
      \\"type\\": \\"image/x-icon\\"
    },
    {
      \\"src\\": \\"logo192.png\\",
      \\"type\\": \\"image/png\\",
      \\"sizes\\": \\"192x192\\"
    },
    {
      \\"src\\": \\"logo512.png\\",
      \\"type\\": \\"image/png\\",
      \\"sizes\\": \\"512x512\\"
    }
  ],
  \\"start_url\\": \\".\\",
  \\"display\\": \\"standalone\\",
  \\"theme_color\\": \\"#000000\\",
  \\"background_color\\": \\"#ffffff\\"
}
"
`;

exports[`when working with an app can generate a asset-manifest 1`] = `
"{
  \\"files\\": {
    \\"main.css\\": \\"/static/css/main.f09a1402.chunk.css\\",
    \\"main.js\\": \\"/static/js/main.b20ef164.chunk.js\\",
    \\"main.js.map\\": \\"/static/js/main.b20ef164.chunk.js.map\\",
    \\"runtime-main.js\\": \\"/static/js/runtime-main.46d4e883.js\\",
    \\"runtime-main.js.map\\": \\"/static/js/runtime-main.46d4e883.js.map\\",
    \\"static/js/2.9da65802.chunk.js\\": \\"/static/js/2.9da65802.chunk.js\\",
    \\"static/js/2.9da65802.chunk.js.map\\": \\"/static/js/2.9da65802.chunk.js.map\\",
    \\"index.html\\": \\"/index.html\\",
    \\"static/css/main.f09a1402.chunk.css.map\\": \\"/static/css/main.f09a1402.chunk.css.map\\",
    \\"static/js/2.9da65802.chunk.js.LICENSE.txt\\": \\"/static/js/2.9da65802.chunk.js.LICENSE.txt\\"
  },
  \\"entrypoints\\": [
    \\"static/js/runtime-main.46d4e883.js\\",
    \\"static/js/2.9da65802.chunk.js\\",
    \\"static/css/main.f09a1402.chunk.css\\",
    \\"static/js/main.b20ef164.chunk.js\\"
  ]
}"
`;

exports[`when working with an app can generate a index.html 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\" />
    <link rel=\\"icon\\" href=\\"/favicon.ico\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"theme-color\\" content=\\"#000000\\" />
    <meta
      name=\\"description\\"
      content=\\"Web site created using create-react-app\\"
    />
    <link rel=\\"apple-touch-icon\\" href=\\"/logo192.png\\" />
    <link rel=\\"manifest\\" href=\\"/manifest.json\\" />
    <title>React App</title>
    <link href=\\"/static/css/main.f09a1402.chunk.css\\" rel=\\"stylesheet\\" />
  </head>
  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id=\\"root\\"></div>
    <script>
      !(function (e) {
        function r(r) {
          for (
            var n, p, l = r[0], a = r[1], f = r[2], c = 0, s = [];
            c < l.length;
            c++
          )
            (p = l[c]),
              Object.prototype.hasOwnProperty.call(o, p) &&
                o[p] &&
                s.push(o[p][0]),
              (o[p] = 0);
          for (n in a)
            Object.prototype.hasOwnProperty.call(a, n) && (e[n] = a[n]);
          for (i && i(r); s.length; ) s.shift()();
          return u.push.apply(u, f || []), t();
        }
        function t() {
          for (var e, r = 0; r < u.length; r++) {
            for (var t = u[r], n = !0, l = 1; l < t.length; l++) {
              var a = t[l];
              0 !== o[a] && (n = !1);
            }
            n && (u.splice(r--, 1), (e = p((p.s = t[0]))));
          }
          return e;
        }
        var n = {},
          o = { 1: 0 },
          u = [];
        function p(r) {
          if (n[r]) return n[r].exports;
          var t = (n[r] = { i: r, l: !1, exports: {} });
          return e[r].call(t.exports, t, t.exports, p), (t.l = !0), t.exports;
        }
        (p.m = e),
          (p.c = n),
          (p.d = function (e, r, t) {
            p.o(e, r) ||
              Object.defineProperty(e, r, { enumerable: !0, get: t });
          }),
          (p.r = function (e) {
            \\"undefined\\" != typeof Symbol &&
              Symbol.toStringTag &&
              Object.defineProperty(e, Symbol.toStringTag, { value: \\"Module\\" }),
              Object.defineProperty(e, \\"__esModule\\", { value: !0 });
          }),
          (p.t = function (e, r) {
            if ((1 & r && (e = p(e)), 8 & r)) return e;
            if (4 & r && \\"object\\" == typeof e && e && e.__esModule) return e;
            var t = Object.create(null);
            if (
              (p.r(t),
              Object.defineProperty(t, \\"default\\", { enumerable: !0, value: e }),
              2 & r && \\"string\\" != typeof e)
            )
              for (var n in e)
                p.d(
                  t,
                  n,
                  function (r) {
                    return e[r];
                  }.bind(null, n)
                );
            return t;
          }),
          (p.n = function (e) {
            var r =
              e && e.__esModule
                ? function () {
                    return e.default;
                  }
                : function () {
                    return e;
                  };
            return p.d(r, \\"a\\", r), r;
          }),
          (p.o = function (e, r) {
            return Object.prototype.hasOwnProperty.call(e, r);
          }),
          (p.p = \\"/\\");
        var l = (this[\\"webpackJsonpsample-app\\"] =
            this[\\"webpackJsonpsample-app\\"] || []),
          a = l.push.bind(l);
        (l.push = r), (l = l.slice());
        for (var f = 0; f < l.length; f++) r(l[f]);
        var i = a;
        t();
      })([]);
    </script>
    <script src=\\"/static/js/2.9da65802.chunk.js\\"></script>
    <script src=\\"/static/js/main.b20ef164.chunk.js\\"></script>
  </body>
</html>
"
`;

exports[`when working with an app can generate a manifest 1`] = `
"{
  \\"short_name\\": \\"React App\\",
  \\"name\\": \\"Create React App Sample\\",
  \\"icons\\": [
    {
      \\"src\\": \\"favicon.ico\\",
      \\"sizes\\": \\"64x64 32x32 24x24 16x16\\",
      \\"type\\": \\"image/x-icon\\"
    },
    {
      \\"src\\": \\"logo192.png\\",
      \\"type\\": \\"image/png\\",
      \\"sizes\\": \\"192x192\\"
    },
    {
      \\"src\\": \\"logo512.png\\",
      \\"type\\": \\"image/png\\",
      \\"sizes\\": \\"512x512\\"
    }
  ],
  \\"start_url\\": \\".\\",
  \\"display\\": \\"standalone\\",
  \\"theme_color\\": \\"#000000\\",
  \\"background_color\\": \\"#ffffff\\"
}
"
`;
